File: Data Structures & Algorithms/majority-element/submission-7.cs
public class Solution {
    public int MajorityElement(int[] nums) {
        // Dictionary <int, int> hMap = new Dictionary <int, int>();
        // int n = nums.Length;
        // foreach(int num in nums){
        //     if(hMap.ContainsKey(num))
        //         hMap[num]++;
        //     else
        //         hMap[num]=1;
        // }
        // foreach(KeyValuePair<int, int> pair in hMap){
        //     if(pair.Value>(n/2))
        //         return pair.Key;
        // }
        // return -1;

        int ele = 0;
        int cnt = 0;
        foreach(int num in nums){
            if(cnt==0)
                ele=num;
            if(ele==num)
                cnt++;
            else
                cnt--;
        }
        return ele;
    }
}

File: Data Structures & Algorithms/top-k-elements-in-list/submission-7.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # Sorting + HashMap
        # hMap={}
        # for num in nums:
        #     hMap[num]=1+hMap.get(num,0)
        
        # sorted_freq = sorted(hMap.items(), key=lambda x: x[1], reverse=True)
        # return [num for num, cnt in sorted_freq[:k]]

        # MIN HEAP
        # hMap = {}
        # heap=[]
        # for num in nums:
        #     hMap[num]=1+hMap.get(num,0)
        # for num , cnt in hMap.items():
        #     heapq.heappush(heap, (cnt , num))
        #     if len(heap)>k:
        #         heapq.heappop(heap)
        # return [num for cnt, num in heap]
        
        # BUCKET SORT
        hMap = {}
        for num in nums:
            hMap[num]=1+hMap.get(num,0)
        # create bucket
        bucket = [[] for _ in range(len(nums)+1)]

        for num, cnt in hMap.items():
            bucket[cnt].append(num)
        
        res=[]
        for i in range(len(bucket)-1, 0, -1):
            for n in bucket[i]:
                res.append(n)
                if len(res)==k:
                    return res






